File: src/commands/nick.ts
import { Command } from 'commander';
import { getContext, handleError } from './shared.js';
import { updateProjectConfig, type ProjectConfig } from '../db/jsonl.js';
import { normalizeAgentRef, isValidAgentId } from '../core/agents.js';

type KnownAgentEntry = NonNullable<ProjectConfig['known_agents']>[string];

function findKnownAgent(config: ProjectConfig | null, ref: string): { guid: string; entry: KnownAgentEntry } | null {
  if (!config?.known_agents) return null;
  const normalized = normalizeAgentRef(ref);

  if (config.known_agents[normalized]) {
    return { guid: normalized, entry: config.known_agents[normalized] };
  }

  for (const [guid, entry] of Object.entries(config.known_agents)) {
    if (entry.name === normalized || entry.global_name === normalized) {
      return { guid, entry };
    }
    if (entry.nicks && entry.nicks.includes(normalized)) {
      return { guid, entry };
    }
  }

  return null;
}

export function nickCommand(): Command {
  return new Command('nick')
    .description('Add a nickname for an agent in this channel')
    .argument('<agent>', 'agent name or GUID')
    .requiredOption('--as <nickname>', 'nickname to add')
    .action((agentRef: string, options, cmd) => {
      try {
        const { project, jsonMode, projectConfig } = getContext(cmd);
        const nickname = normalizeAgentRef(options.as);

        if (!isValidAgentId(nickname)) {
          throw new Error(`Invalid nickname: ${nickname}`);
        }

        const found = findKnownAgent(projectConfig ?? null, agentRef);
        if (!found) {
          throw new Error(`Agent not found in known_agents: ${agentRef}`);
        }

        const existing = found.entry.nicks ?? [];
        const next = existing.includes(nickname) ? existing : [...existing, nickname];
        updateProjectConfig(project.dbPath, {
          known_agents: {
            [found.guid]: {
              nicks: next,
            },
          },
        });

        if (jsonMode) {
          console.log(JSON.stringify({
            agent_id: found.guid,
            name: found.entry.name,
            nicks: next,
          }, null, 2));
        } else {
          console.log(`Added nickname @${nickname} for @${found.entry.name}`);
        }
      } catch (error) {
        handleError(error);
      }
    });
}

export function nicksCommand(): Command {
  return new Command('nicks')
    .description('Show nicknames for an agent in this channel')
    .argument('<agent>', 'agent name or GUID')
    .action((agentRef: string, options, cmd) => {
      try {
        const { jsonMode, projectConfig } = getContext(cmd);
        const found = findKnownAgent(projectConfig ?? null, agentRef);
        if (!found) {
          throw new Error(`Agent not found in known_agents: ${agentRef}`);
        }

        const nicks = found.entry.nicks ?? [];
        if (jsonMode) {
          console.log(JSON.stringify({
            agent_id: found.guid,
            name: found.entry.name,
            global_name: found.entry.global_name ?? null,
            nicks,
          }, null, 2));
        } else {
          console.log(`@${found.entry.name} (${found.guid})`);
          if (nicks.length === 0) {
            console.log('  (no nicknames)');
          } else {
            for (const nick of nicks) {
              console.log(`  @${nick}`);
            }
          }
        }
      } catch (error) {
        handleError(error);
      }
    });
}

export function whoamiCommand(): Command {
  return new Command('whoami')
    .description('Show your known names and nicknames')
    .action((options, cmd) => {
      try {
        const { jsonMode, projectConfig } = getContext(cmd);
        const envAgent = process.env.MM_AGENT_ID;
        if (!envAgent) {
          throw new Error('MM_AGENT_ID not set. Run `mm new` or `mm back` first.');
        }

        const found = findKnownAgent(projectConfig ?? null, envAgent);
        if (!found) {
          throw new Error(`Agent not found in known_agents: ${envAgent}`);
        }

        const nicks = found.entry.nicks ?? [];
        const payload = {
          agent_id: found.guid,
          name: found.entry.name,
          global_name: found.entry.global_name ?? null,
          nicks,
        };

        if (jsonMode) {
          console.log(JSON.stringify(payload, null, 2));
        } else {
          console.log(`You are @${found.entry.name} (${found.guid})`);
          if (payload.global_name) {
            console.log(`  global: @${payload.global_name}`);
          }
          if (nicks.length === 0) {
            console.log('  nicknames: (none)');
          } else {
            console.log('  nicknames:');
            for (const nick of nicks) {
              console.log(`    @${nick}`);
            }
          }
        }
      } catch (error) {
        handleError(error);
      }
    });
}

File: internal/command/nick.go
package command

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/adamavenir/fray/internal/core"
	"github.com/adamavenir/fray/internal/db"
	"github.com/spf13/cobra"
)

// NewNickCmd creates the nick command.
func NewNickCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "nick <agent>",
		Short: "Add a nickname for an agent in this channel",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			ctx, err := GetContext(cmd)
			if err != nil {
				return writeCommandError(cmd, err)
			}
			defer ctx.DB.Close()

			nick, _ := cmd.Flags().GetString("as")
			nick = core.NormalizeAgentRef(nick)
			if nick == "" {
				return writeCommandError(cmd, fmt.Errorf("--as is required"))
			}
			if !core.IsValidAgentID(nick) {
				return writeCommandError(cmd, fmt.Errorf("invalid nickname: %s", nick))
			}

			found := findKnownAgent(ctx.ProjectConfig, args[0])
			if found == nil {
				return writeCommandError(cmd, fmt.Errorf("agent not found in known_agents: %s", args[0]))
			}

			existing := found.Entry.Nicks
			next := existing
			if !containsString(existing, nick) {
				next = append(append([]string{}, existing...), nick)
			}

			update := db.ProjectConfig{
				KnownAgents: map[string]db.ProjectKnownAgent{
					found.GUID: {Nicks: next},
				},
			}
			if _, err := db.UpdateProjectConfig(ctx.Project.DBPath, update); err != nil {
				return writeCommandError(cmd, err)
			}

			if ctx.JSONMode {
				payload := map[string]any{
					"agent_id": found.GUID,
					"name":     derefName(found.Entry.Name, ""),
					"nicks":    next,
				}
				return json.NewEncoder(cmd.OutOrStdout()).Encode(payload)
			}

			name := derefName(found.Entry.Name, found.GUID)
			fmt.Fprintf(cmd.OutOrStdout(), "Added nickname @%s for @%s\n", nick, name)
			return nil
		},
	}

	cmd.Flags().String("as", "", "nickname to add")
	_ = cmd.MarkFlagRequired("as")
	return cmd
}

// NewNicksCmd creates the nicks command.
func NewNicksCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "nicks <agent>",
		Short: "Show nicknames for an agent in this channel",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			ctx, err := GetContext(cmd)
			if err != nil {
				return writeCommandError(cmd, err)
			}
			defer ctx.DB.Close()

			found := findKnownAgent(ctx.ProjectConfig, args[0])
			if found == nil {
				return writeCommandError(cmd, fmt.Errorf("agent not found in known_agents: %s", args[0]))
			}

			nicks := found.Entry.Nicks
			payload := map[string]any{
				"agent_id":    found.GUID,
				"name":        derefName(found.Entry.Name, ""),
				"global_name": derefName(found.Entry.GlobalName, ""),
				"nicks":       nicks,
			}

			if ctx.JSONMode {
				return json.NewEncoder(cmd.OutOrStdout()).Encode(payload)
			}

			name := derefName(found.Entry.Name, found.GUID)
			fmt.Fprintf(cmd.OutOrStdout(), "@%s (%s)\n", name, found.GUID)
			if len(nicks) == 0 {
				fmt.Fprintln(cmd.OutOrStdout(), "  (no nicknames)")
				return nil
			}
			for _, nick := range nicks {
				fmt.Fprintf(cmd.OutOrStdout(), "  @%s\n", nick)
			}
			return nil
		},
	}

	return cmd
}

// NewWhoamiCmd creates the whoami command.
func NewWhoamiCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "whoami",
		Short: "Show your known names and nicknames",
		RunE: func(cmd *cobra.Command, args []string) error {
			ctx, err := GetContext(cmd)
			if err != nil {
				return writeCommandError(cmd, err)
			}
			defer ctx.DB.Close()

			envAgent := os.Getenv("FRAY_AGENT_ID")
			if envAgent == "" {
				return writeCommandError(cmd, fmt.Errorf("FRAY_AGENT_ID not set. Run `fray new` or `fray back` first."))
			}

			found := findKnownAgent(ctx.ProjectConfig, envAgent)
			if found == nil {
				return writeCommandError(cmd, fmt.Errorf("agent not found in known_agents: %s", envAgent))
			}

			nicks := found.Entry.Nicks
			payload := map[string]any{
				"agent_id":    found.GUID,
				"name":        derefName(found.Entry.Name, ""),
				"global_name": derefName(found.Entry.GlobalName, ""),
				"nicks":       nicks,
			}

			if ctx.JSONMode {
				return json.NewEncoder(cmd.OutOrStdout()).Encode(payload)
			}

			name := derefName(found.Entry.Name, found.GUID)
			fmt.Fprintf(cmd.OutOrStdout(), "You are @%s (%s)\n", name, found.GUID)
			globalName := derefName(found.Entry.GlobalName, "")
			if globalName != "" {
				fmt.Fprintf(cmd.OutOrStdout(), "  global: @%s\n", globalName)
			}
			if len(nicks) == 0 {
				fmt.Fprintln(cmd.OutOrStdout(), "  nicknames: (none)")
				return nil
			}
			fmt.Fprintln(cmd.OutOrStdout(), "  nicknames:")
			for _, nick := range nicks {
				fmt.Fprintf(cmd.OutOrStdout(), "    @%s\n", nick)
			}
			return nil
		},
	}

	return cmd
}

func containsString(values []string, value string) bool {
	for _, item := range values {
		if item == value {
			return true
		}
	}
	return false
}
